File: backend/src/main/java/com/nischal/backend/service/BillingService.java
package com.nischal.backend.service;

import com.nischal.backend.dto.billing.*;

public interface BillingService {
    BillResponse getBill(Long orderId);
    BillResponse processPayment(Long orderId, ProcessPaymentRequest request, String performedBy);
    SplitBillResponse splitBill(Long orderId, SplitBillRequest request);

    /** Initiates a Khalti payment for the order and returns pidx + payment URL. */
    KhaltiInitiateResponse initiateKhaltiPayment(Long orderId, String returnUrl);

    /** Verifies a Khalti payment by pidx; marks order PAID if status is Completed. */
    BillResponse verifyKhaltiPayment(KhaltiVerifyRequest request);

    /** Customer requests to pay with cash — notifies staff via WebSocket. */
    void requestCashPayment(Long orderId);

    /** Staff confirms cash was received — marks order PAID and awards loyalty points. */
    BillResponse confirmCashPayment(Long orderId, String performedBy);
}

File: frontend/src/services/billing.service.ts
import { apiClient } from './api.service';
import type {
  BillResponse,
  ProcessPaymentRequest,
  SplitBillRequest,
  SplitBillResponse,
  KhaltiInitiateResponse,
  KhaltiVerifyRequest,
} from '../types/billing.types';

export const billingService = {
  async getBill(orderId: number): Promise<BillResponse> {
    return apiClient.get<BillResponse>(`/billing/${orderId}`);
  },

  async processPayment(orderId: number, data: ProcessPaymentRequest): Promise<BillResponse> {
    return apiClient.post<BillResponse>(`/billing/${orderId}/pay`, data);
  },

  async splitBill(orderId: number, data: SplitBillRequest): Promise<SplitBillResponse> {
    return apiClient.post<SplitBillResponse>(`/billing/${orderId}/split`, data);
  },

  async initiateKhaltiPayment(orderId: number, returnUrl?: string): Promise<KhaltiInitiateResponse> {
    const params = returnUrl ? `?returnUrl=${encodeURIComponent(returnUrl)}` : '';
    return apiClient.post<KhaltiInitiateResponse>(`/billing/${orderId}/khalti/initiate${params}`);
  },

  /** Public — called by customer directly from order tracker (no staff auth needed) */
  async customerInitiateKhaltiPayment(orderId: number): Promise<KhaltiInitiateResponse> {
    return apiClient.post<KhaltiInitiateResponse>(`/billing/${orderId}/khalti/pay`);
  },

  async verifyKhaltiPayment(data: KhaltiVerifyRequest): Promise<BillResponse> {
    return apiClient.post<BillResponse>('/billing/khalti/verify', data);
  },

  /** Public — customer signals they want to pay with cash; staff must confirm */
  async requestCashPayment(orderId: number): Promise<void> {
    return apiClient.post<void>(`/billing/${orderId}/cash/request`);
  },

  /** Staff only — confirms cash was received, marks PAID, awards loyalty points */
  async confirmCashPayment(orderId: number): Promise<BillResponse> {
    return apiClient.post<BillResponse>(`/billing/${orderId}/cash/confirm`);
  },
};
